src/m_class.c

```diff
@@ -933,24 +933,194 @@ t_pd *pd_newest(void)
 
     /* horribly, we need prototypes for each of the artificial function
     calls in typedmess(), to keep the compiler quiet. */
+    /* Even more horribly, function pointers must be called at their
+    exact type in Emscripten, including return type, which makes the
+    number of cases explode (7 * 6 * 2 = 84). This is not checked at
+    compile time, but wrong usage crashes at runtime. The naming pattern
+    is t_funMN for M t_int args and N t_floatarg args returning t_pd *,
+    with t_vfunMN for the void-returning variants. The versions with a
+    t_pd * are for pd_objectmaker, which returns the new object from the
+    method, c.f. pd_newest support. For the same reason, bang_new() and
+    similar can't be private in Emscripten, because their return type is
+    different (as pd_objectmaker methods) vs regular bang methods, so
+    they need special treatment in the dispatcher: the dispatcher checks
+    explicitly for pd_objectmaker and calls the method at the correct
+    type. */
 typedef t_pd *(*t_newgimme)(t_symbol *s, int argc, t_atom *argv);
 typedef void(*t_messgimme)(t_pd *x, t_symbol *s, int argc, t_atom *argv);
+typedef void*(*t_messgimmer)(t_pd *x, t_symbol *s, int argc, t_atom *argv);
+
+typedef t_pd *(*t_fun00)(void);
+typedef t_pd *(*t_fun10)(t_int i1    );
+typedef t_pd *(*t_fun20)(t_int i1, t_int i2    );
+typedef t_pd *(*t_fun30)(t_int i1, t_int i2, t_int i3    );
+typedef t_pd *(*t_fun40)(t_int i1, t_int i2, t_int i3, t_int i4    );
+typedef t_pd *(*t_fun50)(t_int i1, t_int i2, t_int i3, t_int i4, t_int i5    );
+typedef t_pd *(*t_fun60)(t_int i1, t_int i2, t_int i3, t_int i4, t_int i5, t_int i6    );
+
+typedef t_pd *(*t_fun01)(
+    t_floatarg d1);
+typedef t_pd *(*t_fun11)(t_int i1,
+    t_floatarg d1);
+typedef t_pd *(*t_fun21)(t_int i1, t_int i2,
+    t_floatarg d1);
+typedef t_pd *(*t_fun31)(t_int i1, t_int i2, t_int i3,
+    t_floatarg d1);
+typedef t_pd *(*t_fun41)(t_int i1, t_int i2, t_int i3, t_int i4,
+    t_floatarg d1);
+typedef t_pd *(*t_fun51)(t_int i1, t_int i2, t_int i3, t_int i4, t_int i5,
+    t_floatarg d1);
+typedef t_pd *(*t_fun61)(t_int i1, t_int i2, t_int i3, t_int i4, t_int i5, t_int i6,
+    t_floatarg d1);
+
```

Comment:
Some of those function prototypes (and the corresponding case labels in `pd_typedmess`) are redundant. Pd only allows up to 5 typed arguments, so we don't need `t_fun53`, for example.

Also, why do we need `t_fun6*` or `t_fun*6` at all if there are only max. 5 arguments? The original implementation also did this, but I don't really understand why. Can you figure it out and add an explanation? :-)

Reply:
And just in case anyone worries: having 21 case labels instead of 6 doesn't make a difference in performance because the compiler turns the switch statement into a jump table. Here's my version of the code with the assembly output:
https://godbolt.org/z/49vbq3x13

GCC produces a jump table, as expected.

Clang, however, does some crazy optimizations which I don't 100% understand. I *think* it recognizes that all those function signatures only differ in the number of integer resp. float arguments. With the System V AMD64 ABI, the first 6 integer/pointer arguments and the first 8 floating point arguments are passed in registers (https://en.wikipedia.org/wiki/X86_calling_conventions#System_V_AMD64_ABI). This means we can simply iterate over the `t_atom` argument vector, push the values to the appropriate registers and then call the function pointer *unconditionally*!

BTW, I encode the number of symbol/gpointer and float arguments in hex instead of decimal, because it is a bit more efficient and also more elegant :-)